### PR TITLE
[WIP #93] - Revise add-dimension

### LINKED ISSUE #93: Revise add_dimension
Currently we make the dimension even by adding a dummy dimension. However, we can now approach odd dimensions naturally. So we should remove this part and instead only check that the `partition` is consistent with the dimension.

### PR DESCRIPTION
**Pull request by mforets**
Closes #93.

### CODE REVIEW

**Comment by mforets** [comment]
There is some conflict with `partition` option that i didn't look into yet.

```julia
julia> s = solve(ContinuousSystem(A, X0), :T=>0.1, :partition=>[1:2, 3:4, 5],
                 :block_types=>Dict(HPolygon=>[1:2, 3:4], Interval=>[[5]]),
                 :vars=>[1,3]);
ERROR: Option :partition must be of 'AbstractArray{#s60,1} where #s60<:AbstractArray{Int64,1}' type.
Stacktrace:
 [1] validate_solver_options_and_add_default_values!(::Reachability.Options) at /Users/forets/.julia/v0.6/Reachability/src/options.jl:276
 [2] solve(::Reachability.Systems.ContinuousSystem, ::Reachability.Options) at /Users/forets/.julia/v0.6/Reachability/src/solve.jl:92
 [3] solve(::Reachability.Systems.ContinuousSystem, ::Pair{Symbol,Float64}, ::Pair{Symbol,Array{Any,1}}, ::Vararg{Pair{Symbol,#s61} where #s61,N} where N) at /Users/forets/.julia/v0.6/Reachability/src/solve.jl:23
4
```

File: test/Reachability/unit_solve.jl [hunk 1]
```diff
@@ -47,3 +47,12 @@ s = solve(ContinuousSystem(sparse(A), X0), :T=>0.1, :partition=>[1:2, 3:4],
 
 s = solve(ContinuousSystem(sparse(A), X0), :T=>0.1, :partition=>[[i] for i in 1:4],
           :set_type=>Interval, :vars=>[1,3], :lazy_expm=>true);
+
+# ===============================
+# System with an add dimension
```
Comment (schillic) [inline test/Reachability/unit_solve.jl]: "odd"

File: src/options.jl [hunk 2]
```diff
@@ -478,11 +478,14 @@ function check_and_add_partition_block_types!(dict::Dict{Symbol,Any},
...
 
-    block_types = haskey(dict, :block_types) ? dict[:block_types] :
-        haskey(dict, :set_type) ?
-            Dict{Type{<:LazySet}, AbstractVector{<:AbstractVector{Int}}}(
-                dict[:set_type] => copy(dict_copy[:partition])
-            ) : nothing
+    block_types = nothing
+    if haskey(dict, :block_types)
+        block_types = convert(Dict{Type{<:LazySet}, AbstractVector{<:AbstractVector{Int}}}, dict[:block_types])
+    else
```
Comment (schillic) [inline src/options.jl]: merge `else` and `if` to `elseif`?

**Comment by schillic** [comment]
> There is some conflict with partition option that i didn't look into yet.

The problem is that `:partition` has the value `Any[1:2, 3:4, 5]`. It must be of type `AbstractVector{<:AbstractVector{Int}}`, not `AbstractVector{<:AbstractVector{Any}}`.

File: src/options.jl [hunk 3]
```diff
@@ -206,8 +206,8 @@ function validate_solver_options_and_add_default_values!(options::Options)::Opti
         elseif key == :partition
             expected_type = AbstractVector{<:AbstractVector{Int}}
         elseif key == :block_types
-            expected_type =
-                Dict{Type{<:LazySet}, AbstractVector{<:AbstractVector{Int}}}
+            expected_type = Dict
+                #Dict{Type{<:LazySet}, AbstractVector{<:AbstractVector{Int}}}
```
Comment (schillic) [inline src/options.jl]: I think you can re-add this because you `convert` the value to this type.

**Comment by schillic** [comment]
> thanks. i am still getting

We have `bi = 1:2`, `bj = 5` in `arr[j] = ϕpowerk[bi, bj] * Xhat0[j]` from `reach_blocks.jl`.

The problem is this. In `LazySets.LinearMap` we have:
```julia
function *(a::N, X::LazySet) where {N}
    return LinearMap(a * speye(N, dim(X)), X)
end
```
We call it with the following:
```julia
X = Interval([0.9, 1.10445]); a = [0.0029649, -0.00173048];
a * X
```
Here `N` becomes `Array{Float64,2}`, and there is no `speye`(N, k)`.

**Pull request merged**